jctools-core/src/main/java/org/jctools/queues/MpmcArrayQueue.java

```diff
@@ -273,16 +273,30 @@ public E poll()
     public E peek()
     {
         long cIndex;
+        long pIndex;
         E e;
         do
         {
+            // Order matters!
             cIndex = lvConsumerIndex();
-            // other consumers may have grabbed the element, or queue might be empty
-            e = lpRefElement(buffer, calcCircularRefElementOffset(cIndex, mask));
-            // only return null if queue is empty
+            // Volatile Mode: Loading element must be completed before loading pIndex.
+            e = lvRefElement(buffer, calcCircularRefElementOffset(cIndex, mask));
+            pIndex = lvProducerIndex();
```

Comment:
The original solution relying on the matching sequence is more in line with how `poll` works and the spirit of this queue altogether. The idea here is to avoid as much as possible cache misses/traffic induced by loading the producer index from the consumer threads.

Reply:
The sequence could be read twice to check if the read element is stable, similarly to what a stamped lock does

Reply:
Stable value is unnecessary